src/Discord.hs

```diff
@@ -117,14 +118,18 @@ data RestCallErrorCode = RestCallErrorCode Int T.Text T.Text
   deriving (Show, Eq, Ord)
 
 -- | Execute one http request and get a response
-restCall :: (FromJSON a, Request (r a)) => DiscordHandle -> r a -> IO (Either RestCallErrorCode a)
+restCall :: (FromJSON (Response r), Request r) => DiscordHandle -> r -> IO (Either RestCallErrorCode (Response r))
 restCall h r = do empty <- isEmptyMVar (discordHandleLibraryError h)
                   if not empty
                   then pure (Left (RestCallErrorCode 400 "Library Stopped Working" ""))
                   else do
                       resp <- writeRestCall (discordHandleRestChan h) r
                       case resp of
-                        Right x -> pure (Right x)
+                        Right x -> do
+                            c <- readCache h
+                            let c' = updateCache c r x
+                            putMVar (snd $ discordHandleCache h) (Right c')
```

Comment:
This code is the cause of the slowness/crashes. The cache is stored in an `MVar`, which is a thread communication device which can store a single object. `readCache` calls `readMVar`, which is going to return a copy of the cache. The `MVar` remains full. Then `putMVar` is going to try and put `c` into the `MVar`, and will block until the `MVar` is empty. 

The cache is kept inside an `MVar` so many threads can safely modify it at once. Get the cache with `takeMVar`, modify it, then put it back with `putMVar`.

Reply:
Have you tried switching to STM? It handles concurrency quite well and isn't a bit leap from MVars